### PR TITLE
Yuhsuan/documentation

### LINKED ISSUE #1482: Annotate low-level scripting actions in various stores with JSDoc
Based on request in https://github.com/CARTAvis/carta/issues/78 from @keflavich.

We need to have some form of documentation for the low-level scripting actions that are available. This will allow users to either utilise the debug execution feature on the frontend, or extend the high-level Python scripting interface. 

I suggest we use JSDoc for this, as we can embed the documentation in the code and generate an API document in HTML reasonably easily.

As an example, the [angus/jsdoc branch](https://github.com/CARTAvis/carta-frontend/tree/angus/jsdoc) has the following for `appendFile`: 

```
/**     
* @access public
* @async
* Appends a file at the given path to the list of existing open files
* @param path - path to the parent directory of the file to open, or of the file itself     
* @param {string=} filename - filename of the file to open     
* @param {string=} hdu - HDU to open. If left blank, the first image HDU will be opened     
* @return {Promise<number>} [async] the file ID of the opened file     
* */
* 
@action appendFile = (path: string, filename: string, hdu: string) => {
...
```

If we can document these functions well, and also extend the debug execution dialog into something more akin to a full scripting console, CARTA would become very extensible.

### LINKED ISSUE #1556: Additional tutorial entries for code snippets
Additional tutorial entries are needed for the following:

- [ ] Changing FoV, channels and Stokes of images
- [ ] Adjusting render config
- [ ] Creating regions

@kswang1029 may have more suggestions on what to add after testing out the code snippets feature in more detail

### LINKED ISSUE #1873: create document for developer guidelines
create a `contributing.md` document for developer guidelines including:
* filing issues
* sending pull requests
* coding conventions
* labeling conventions
* ...

### PR DESCRIPTION
**Pull request by YuHsuan-Hwang**
**Description**

1. Closes #1482.

* Updated the documentation in the codebase from JSDoc to TSDoc.
* Added eslint plugin `eslint-plugin-tsdoc` to check for the required TSDoc format. Incorrect format will result in a compile error.
  To run lint checks:
  ```
  npm run check-eslint
  ```
* Implemented a documentation website using [Docusaurus](https://docusaurus.io).
  To build the website locally:
  ```
  cd docs_website
  npm install
  npm start
  ```
  To create a production build locally:
  ```
  npm run build
  npm run serve
  ```

  The website includes:

    * carta-frontend API: API pages are automatically generated from the codebase TSDoc documentation with [docusaurus-plugin-typedoc-api](https://github.com/milesj/docusaurus-plugin-typedoc-api). See `documentation-guidelines.md` “Writing API documentation” for details.
    * Documentation pages: Additional content is generated from markdown files. See `documentation-guidelines.md` “Writing documentation pages” for details.
    * Searching: The search functionality is implemented using [docusaurus-lunr-search](https://github.com/praveenn77/docusaurus-lunr-search). The feature is only available in production builds. Note that files in `theme/SearchBar` are automatically generated by running `npm run swizzle docusaurus-lunr-search SearchBar -- --eject --danger`.
    * Prettier formatting: Files can be reformatted by `npm run reformat`, including [markdown files](https://prettier.io/blog/2017/11/07/1.8.0.html#markdown-support).
   
  Notes:

    * The root dir is named as `docs_website/` instead of `docs/`. This is because Docusaurus requires a sub dir `docs/`, and `docs/docs/` is confusing.
    * Node v16.14 or above is required. Tried that the build fails with node v14.18.
* Added CI checks for formatting and deployment of the documentation website. Used cloud service because our servers will be busy with `node-v16` and `node-v18` tasks.
* Added Github actions for updating GitHub Pages whenever the dev branch is updated. To be tested after merging the PR.

2. Closes #1556.
  
    Added code snippet tutorials at `carta-frontend/docs/category/code-snippet-tutorial` in the documentation website.
    The first two tutorials are revised from the example code snippets in the codebase. Added some new content such as image properties and regions. Please review the content of the pages.

3. Closes #1873.
  
    Added developer guidelines at `carta-frontend/docs/category/contributing` in the documentation website, including topics of developer tips, Github workflow, unit test guidelines, and documentation guidelines. Please review the content of the pages.

Other changes:
* Updated example snippets for consistency.
* Use CRA built-in ESLint
* Fixed unit test ESLint error

Action items after merging the PR:
* Setup GitHub Pages settings, and test the Github action for GitHub Pages.
* Check if the “Edit this page” links work.
* Check if the tutorial link in the snippets submenu work.

**Checklist**

For linked issues (if there are):
- [ ] assignee and label added
- [ ] ZenHub issue connection, board status, and estimate updated

For the pull request:
- [x] reviewers and assignee added
- [x] ZenHub estimate, milestone, and release (if needed) added
- [x] e2e test passing / ~corresponding fix added~
- [x] changelog updated / ~no changelog update needed~
- [x] ~protobuf updated to the latest dev commit~ / no protobuf update needed
- [x] `BackendService` unchanged / ~`BackendService` changed and corresponding ICD test fix added~

### CODE REVIEW

**Review by ajm-ska — APPROVED** [review]
@YuHsuan-Hwang This is a very impressive looking. Nice work! 
It looks like it should work once merged to the 'dev' branch.

**Review by confluence — APPROVED** [review]
This looks good! A note about deployment: we will need to version this, and have a site for each release as well as the development branch (because users will need to match the API docs to their CARTA release).

**Comment by ajm-ska** [comment]
Here is a static build of the frontend documentation for reviewing: https://carta.asiaa.sinica.edu.tw/carta-frontend

**Comment by kswang1029** [comment]
@YuHsuan-Hwang when I run `npm run check-eslint` I see an error as 

> /Users/kswang/carta_build/carta-frontend/src/components/Shared/ExportImageMenu/ExportImageMenuComponent.test.tsx
>   9:9  error  'mockModifierString' is assigned a value but never used  @typescript-eslint/no-unused-vars
> 
> ✖ 1 problem (1 error, 0 warnings)

do we expect to see this in this branch?

**Comment by kswang1029** [comment]
doing `npm install` in `docs_website` folder results in some WARN. Just in case these are important.

> npm WARN deprecated stable@0.1.8: Modern JS already guarantees Array#sort() is a stable sort, so this library is deprecated. See the compatibility table on MDN: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort#browser_compatibility
> npm WARN deprecated trim@0.0.1: Use String.prototype.trim() instead
> npm WARN deprecated mkdirp@0.3.0: Legacy versions of mkdirp are no longer supported. Please update to mkdirp 1.x. (Note that the API surface has changed to use Promises in 1.x.)
> 
> added 1154 packages, and audited 1155 packages in 17s

**Review by kswang1029 — APPROVED** [review]
Looks great. 👍 

Following are my minor comments:

1. would be good to add links to the API document and code snippets tutorials in the "Snippers sub-menu"
2. in the code snippets tutorial, it would be good to add "loops" and "conditions" to the "Basics" section.

**Review by crocka — APPROVED** [review]

**Comment by YuHsuan-Hwang** [comment]
> when I run `npm run check-eslint` I see an error as
> 
> > /Users/kswang/carta_build/carta-frontend/src/components/Shared/ExportImageMenu/ExportImageMenuComponent.test.tsx
> > 9:9  error  'mockModifierString' is assigned a value but never used  @typescript-eslint/no-unused-vars
> > ✖ 1 problem (1 error, 0 warnings)
> 
> do we expect to see this in this branch?

This is fixed.

> doing `npm install` in `docs_website` folder results in some WARN. Just in case these are important.

These are from dependencies, so we can not fix it.

> 1. would be good to add links to the API document and code snippets tutorials in the "Snippers sub-menu"

Added:
+ Create New Snippet
The link will work after we host it on Github Pages.

> 2. in the code snippets tutorial, it would be good to add "loops" and "conditions" to the "Basics" section.

I don't think we need to add more "JS tutorials" which could be endless ... . I added: "For more usages of ES6-based JavaScript, please refer to the features of the language." in the end of the section.

@kswang1029 Do you think we need to revise the example snippets based on the changes?

**Comment by kswang1029** [comment]
> > when I run `npm run check-eslint` I see an error as
> > > /Users/kswang/carta_build/carta-frontend/src/components/Shared/ExportImageMenu/ExportImageMenuComponent.test.tsx
> > > 9:9  error  'mockModifierString' is assigned a value but never used  @typescript-eslint/no-unused-vars
> > > ✖ 1 problem (1 error, 0 warnings)
> > 
> > 
> > do we expect to see this in this branch?
> 
> This is fixed.
> 
> > doing `npm install` in `docs_website` folder results in some WARN. Just in case these are important.
> 
> These are from dependencies, so we can not fix it.
> 
> > 1. would be good to add links to the API document and code snippets tutorials in the "Snippers sub-menu"
> 
> Added: + Create New Snippet The link will work after we host it on Github Pages.
> 
> > 2. in the code snippets tutorial, it would be good to add "loops" and "conditions" to the "Basics" section.
> 
> I don't think we need to add more "JS tutorials" which could be endless ... . I added: "For more usages of ES6-based JavaScript, please refer to the features of the language." in the end of the section.
> 
> @kswang1029 Do you think we need to revise the example snippets based on the changes?

yes would be good to do so

**Comment by YuHsuan-Hwang** [comment]
Updated example snippets for consistency.

**Review by kswang1029 — APPROVED** [review]

File: docs_website/docs/contributing/github-workflow.md [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+    Check if the end-to-end tests are passing. This will be checked during the review.
+
+-   changelog updated / no changelog update needed:
+
+    Updating the changelog is required when the issue exists in the latest release and the changes have an impact on users. It can be updated when the pull request is ready for merge to prevent frequent merge conflicts.
+
+-   protobuf updated to the latest dev commit / no protobuf update needed:
+
+    Updating the protobuf is required when there are changes in the carat-protobuf submodule. This can be done when the pull request is ready for merge.
```
Comment (TienHao) [inline docs_website/docs/contributing/github-workflow.md]: just found a typo for `carta-protobuf`

Comment (YuHsuan-Hwang) [reply docs_website/docs/contributing/github-workflow.md]: Fixed. Thanks for checking!

**Review by TienHao — APPROVED** [review]

**Pull request merged**